spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTemplate.java

```diff
@@ -0,0 +1,173 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.spanner.core;
+
+import java.util.Arrays;
+import java.util.List;
+
+import com.google.cloud.spanner.DatabaseClient;
+import com.google.cloud.spanner.Key;
+import com.google.cloud.spanner.KeySet;
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.Options;
+import com.google.cloud.spanner.ReadContext;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.Struct;
+
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerMappingContext;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerObjectMapper;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerPersistentEntity;
+import org.springframework.util.Assert;
+
+/**
+ * @author Ray Tsang
+ * @author Chengyuan Zhao
+ */
+public class SpannerTemplate implements SpannerOperations {
+
+	private final DatabaseClient databaseClient;
+
+	private final SpannerMappingContext mappingContext;
+
+	private final SpannerObjectMapper objectMapper;
```

Comment:
Is the assumption that the template will always be used with `@Entity` annotated beans? It then implies that the `SpannerTemplate` cannot be used with JPA.

Reply:
No, the beans don't need to be annotated with that.

Reply:
Hmmm, how do you map columns to bean properties then?

Reply:
Spring Data will map every single property in the given bean type, but it's up to us in SpannerObjectMapper and the mutation factory classes to decide which properties have types that we can handle.

Reply:
What happens when a Spanner column is called "name", and the field in the class is "firstName". How will the user specify the mapping without JPA annotations?

Reply:
For that we have the SpannerColumn annotation that lets you give a different name.

Reply:
I see. We might want to provide other options, for users who don't want to be coupled to Spanner-specific APIs.